x-pack/plugins/rule_registry/server/utils/get_alerts_for_notification.ts

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { MAX_FLAP_COUNT } from '@kbn/alerting-plugin/server/lib/flapping_utils';
+import {
+  ALERT_END,
+  ALERT_STATUS,
+  ALERT_STATUS_ACTIVE,
+  ALERT_STATUS_RECOVERED,
+  EVENT_ACTION,
+} from '@kbn/rule-data-utils';
+
+export function getAlertsForNotification(trackedEventsToIndex: any[]) {
+  return trackedEventsToIndex.map((trackedEvent) => {
+    if (trackedEvent.event[ALERT_STATUS] === ALERT_STATUS_ACTIVE) {
+      trackedEvent.pendingRecoveredCount = 0;
+    } else if (trackedEvent.event[ALERT_STATUS] === ALERT_STATUS_RECOVERED) {
+      if (trackedEvent.flapping) {
+        const count = trackedEvent.pendingRecoveredCount || 0;
+        trackedEvent.pendingRecoveredCount = count + 1;
+        if (trackedEvent.pendingRecoveredCount < MAX_FLAP_COUNT) {
+          trackedEvent.event[ALERT_STATUS] = ALERT_STATUS_ACTIVE;
+          trackedEvent.event[EVENT_ACTION] = 'active';
+          trackedEvent.event[ALERT_END] = {};
```

Comment:
I got an error as below from ruleRegistry  when i testing this PR.
Is it expected?

```
{"took":899,"errors":true,"items":[{"index":{"_index":".internal.alerts-observability.metrics.alerts-default-000001","_id":"850eccca-3f90-4a96-ac65-d2c9d865174c","status":400,"error":{"type":"mapper_parsing_exception","reason":"failed to parse field [kibana.alert.end] of type [date] in document with id '850eccca-3f90-4a96-ac65-d2c9d865174c'. Preview of field's value: '{}'","caused_by":{"type":"illegal_state_exception","reason":"Can't get text on a START_OBJECT at 1:1091"}}}},{"index":{"_index":".internal.alerts-observability.metrics.alerts-default-000001","_id":"820dd6f1-e32d-4238-b634-048703be1008","status":400,"error":{"type":"mapper_parsing_exception","reason":"failed to parse field [kibana.alert.end] of type [date] in document with id '820dd6f1-e32d-4238-b634-048703be1008'. Preview of field's value: '{}'","caused_by":{"type":"illegal_state_exception","reason":"Can't get text on a START_OBJECT at 1:1090"}}}},{"index":{"_index":".internal.alerts-observability.metrics.alerts-default-000001","_id":"70cc5557-2f57-450a-856c-e2715dc7088f","_version":2,"result":"updated","_shards":{"total":1,"successful":1,"failed":0},"_seq_no":14,"_primary_term":1,"status":200}}]}
```

Reply:
No I don't think that's expected, I will look into it!

Reply:
I think it is resolved in this commit b4dff703687ad022a1b4078cca528e1dc74f6a2d
I was setting `ALERT_END` to an empty object when I should have been removing it, I am sorry about that!